File: scripts/validate-workflows.mjs
import { files, fail } from './lib.mjs';
import { readFileSync } from 'node:fs';

const workflows = files('workflows', (name) => name.endsWith('.md'));
const errors = workflows.filter((path) => { const body = readFileSync(path, 'utf8'); return !body.includes('## Flow') || !body.includes('## Controls'); }).map((path) => `WORKFLOW_INVALID: ${path}`);
fail(errors); console.log(`WORKFLOW_VALID: ${workflows.length} workflows`);

File: scripts/validate-workflows.py
#!/usr/bin/env python3
"""Check that workflow documents have a flow and explicit controls."""
import sys
from pathlib import Path


def main(root: str = "workflows") -> int:
    errors = []
    paths = list(Path(root).rglob("*.md"))
    for path in paths:
        text = path.read_text(encoding="utf-8")
        if "## Flow" not in text or "## Controls" not in text:
            errors.append(str(path))
    if errors:
        print("WORKFLOW_VALIDATION_FAILED:\n" + "\n".join(errors))
        return 1
    print(f"WORKFLOW_VALIDATION_PASSED: {len(paths)} workflows")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1] if len(sys.argv) == 2 else "workflows"))
